zyte_spider_templates/spiders/serp.py

```diff
@@ -0,0 +1,128 @@
+from scrapy_spider_metadata import Args
+from w3lib.url import add_or_replace_parameter
+from zyte_common_items import Serp
+
+from ..params import MaxRequestsParam
+from ._google_domains import GoogleDomain
+from .base import BaseSpider
+
+
+class SearchQueriesParam(BaseModel):
+    search_queries: Optional[List[str]] = Field(
+        title="Search Queries",
+        description="Input 1 search query per line (e.g. foo bar).",
+        json_schema_extra={
+            "widget": "textarea",
+        },
+    )
+
+    @field_validator("search_queries", mode="before")
+    @classmethod
+    def validate_search_queries(cls, value: Union[List[str], str]) -> List[str]:
+        """Validate a list of search queries.
+        If a string is received as input, it is split into multiple strings
+        on new lines.
+        """
+        if isinstance(value, str):
+            value = value.split("\n")
+        if not value:
+            raise ValueError("The search_queries parameter value is missing or empty.")
+        result = []
+        for v in value:
+            if not (v := v.strip()):
+                continue
+            result.append(v)
+        return result
+
+
+class SerpMaxPagesParam(BaseModel):
+    max_pages: int = Field(
+        title="Pages",
+        description="Maximum number of result pages to visit per input URL.",
+        default=1,
+    )
+
+
+class GoogleDomainParam(BaseModel):
+    domain: GoogleDomain = Field(
+        title="Domain",
+        description="Target Google domain.",
+        default=GoogleDomain.google_com,
+    )
+
+
+class GoogleSearchSpiderParams(
+    MaxRequestsParam,
+    SerpMaxPagesParam,
+    SearchQueriesParam,
+    GoogleDomainParam,
+    BaseModel,
+):
+    pass
+
+
+class GoogleSearchSpider(Args[GoogleSearchSpiderParams], BaseSpider):
+    """Yield results from Google searches.
+
+    See :class:`~zyte_spider_templates.spiders.serp.GoogleSearchSpiderParams`
+    for supported parameters.
+
+    .. seealso:: :ref:`google-search`.
+    """
+
+    name = "google_search"
+
+    metadata: Dict[str, Any] = {
+        **BaseSpider.metadata,
+        "title": "Google Search Results",
+        "description": "Template for spiders that extract Google search results.",
+    }
+
+    @classmethod
+    def update_settings(cls, settings: BaseSettings) -> None:
+        super().update_settings(settings)
+        retry_policy_setting_priority = settings.getpriority("ZYTE_API_RETRY_POLICY")
+        if (
+            retry_policy_setting_priority is None
+            or retry_policy_setting_priority < SETTINGS_PRIORITIES["spider"]
+        ):
+            settings.set(
+                "ZYTE_API_RETRY_POLICY",
+                "zyte_api.aggressive_retrying",
+                priority="spider",
+            )
+
+    def get_start_request(self, url):
+        return Request(
+            url=url,
+            callback=self.parse_serp,
+            meta={
+                "crawling_logs": {"page_type": "serp"},
```

Comment:
I seems the resulting crawl logs don't quite make sense for serp:

```
Number of Requests per page type:
- product: 0
- nextPage: 0
- subCategories: 0
- productNavigation: 0
- productNavigation-heuristics: 0
- unknown: 0
Structured Logs:
{
  "time": "2024-09-13 12:47:31",
  "current": {
    "url": "https://www.google.com/search?q=frappe",
    "request_url": "https://www.google.com/search?q=frappe",
    "request_fingerprint": "231822665b55ba64d4f531b734de660fe1b81b5d",
    "page_type": "serp",
    "probability": null
  },
  "to_crawl": {
    "product": [],
    "nextPage": [],
    "subCategories": [],
    "productNavigation": [],
    "productNavigation-heuristics": [],
    "unknown": []
  }
}
```

Reply:
Now:

```
Nothing to crawl.
Structured Logs:
{
  "time": "2024-09-17 00:12:01",
  "current": {
    "url": "https://www.google.com/search?q=foo&start=10",
    "request_url": "https://www.google.com/search?q=foo&start=10",
    "request_fingerprint": "1b95b08d33b7b8401ec2c6126fc1399100ab12b9",
    "page_type": "serp",
    "probability": null
  },
  "to_crawl": {}
}
```